ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/modules/cdr/CdrsCpoClient.kt

```diff
@@ -0,0 +1,64 @@
+package com.izivia.ocpi.toolkit.modules.cdr
+
+import com.izivia.ocpi.toolkit.common.*
+import com.izivia.ocpi.toolkit.modules.cdr.domain.Cdr
+import com.izivia.ocpi.toolkit.modules.credentials.repositories.PartnerRepository
+import com.izivia.ocpi.toolkit.modules.versions.domain.ModuleID
+import com.izivia.ocpi.toolkit.transport.TransportClient
+import com.izivia.ocpi.toolkit.transport.TransportClientBuilder
+import com.izivia.ocpi.toolkit.transport.domain.HttpMethod
+import com.izivia.ocpi.toolkit.transport.domain.HttpRequest
+
+/**
+ * Sends calls to an eMSP server
+ * @property transportClientBuilder used to build transport client
+ * @property serverVersionsEndpointUrl used to know which partner to communicate with
+ * @property partnerRepository used to get information about the partner (endpoint, token)
+ */
+class CdrsCpoClient(
+    private val transportClientBuilder: TransportClientBuilder,
+    private val serverVersionsEndpointUrl: String,
+    private val partnerRepository: PartnerRepository
+) : CdrsEmspInterface {
+    private suspend fun buildTransport(): TransportClient = transportClientBuilder
+        .buildFor(
+            module = ModuleID.cdrs,
+            partnerUrl = serverVersionsEndpointUrl,
+            partnerRepository = partnerRepository
+        )
+
+    override suspend fun getCdr(
+        cdrId: CiString
+    ): OcpiResponseBody<Cdr?> =
+        with(buildTransport()) {
+            send(
+                HttpRequest(
+                    method = HttpMethod.GET,
+                    path = "/$cdrId"
+                ).withRequiredHeaders(
+                    requestId = generateRequestId(),
+                    correlationId = generateCorrelationId()
+                )
+                    .authenticate(partnerRepository = partnerRepository, partnerUrl = serverVersionsEndpointUrl)
+            )
+                .parseBody()
+        }
```

Comment:
The documentation states that

> 10.2.2.1. GET Method
Fetch CDRs from the receivers system.
Endpoint structure definition:
No structure defined. This is open to the eMSP to define, the URL is provided to the CPO by the eMSP in the result of the POST request. Therefore, OCPI does not define variables.

So we cannot assume that the eMSP is using that path.

Reply:
I hate the fact that OCPI is a standard, yet it does not specify what should be done in this case... What the point of having a standard if it's up to the parties to decide what should be done 🙃 

Anyway, I think you are right, we cannot assume that `/{cdrId}` will be the endpoint used.

@pbourseau you can remove:

```kotlin
getCdr(cdrId: CiString): OcpiResponseBody<Cdr?>
```

From `CdrsEmspInterface`. The interface will be empty, and `CdrsCpoClient` will only contain `buildTransport` but it's fine.

It will be up to the users of the lib to use their own functions

Reply:
You are right:

> The endpoint returns the requested CDR, if it exists.

We know that the return type will be `OcpiResponseBody<Cdr?>`. So we can define `getCdr` with `url` in parameter

Let's do that, great idea :)